add arm64 builds

martadams89 opened the pull request:
This PR adds arm64 builds back in. 

I forked the repo and tested it against my own and appeared to run as expected.

martadams89 commented:
@acasajus @cquintana92 @nguyenkims can someone review this? It resolves https://github.com/simple-login/app/issues/1566#issuecomment-2463019111

acasajus commented:
We disabled the arm builds because there were some dependencies that were making the tests really flaky and that failed the production deployments. As soon as we fix those dependencies we will enable arm back.

ctrl-i commented:
> We disabled the arm builds because there were some dependencies that were making the tests really flaky and that failed the production deployments. As soon as we fix those dependencies we will enable arm back.

Any progress on this?  If you told us which dependencies were making which test flaky we might be able to help out.  I haven't noticed any flakiness in my testing.

martadams89 commented:
I've just sync'd my fork to the latest code, updated the workflow to mirror the latest in `main` and everything seemed fine - https://github.com/martadams89/sl-app/actions/runs/12846003655

Here's the latest arm64 image - https://registry.hub.docker.com/r/martadams89/sl-app

ctrl-i commented:
> I've just sync'd my fork to the latest code, updated the workflow to mirror the latest in `main` and everything seemed fine - https://github.com/martadams89/sl-app/actions/runs/12846003655
> 
> Here's the latest arm64 image - https://registry.hub.docker.com/r/martadams89/sl-app

@martadams89 Unfortunately this latest build does not seem to work for me.  The database migration service fails with errors in the log of:

```
exec /code/.venv/bin/alembic: exec format error
exec /code/.venv/bin/alembic: exec format error
exec /code/.venv/bin/alembic: exec format error
exec /code/.venv/bin/alembic: exec format error
exec /code/.venv/bin/alembic: exec format error

```
If I disable the migration service, simplelogin-init fails with a single error in the log of:

`exec /code/.venv/bin/python: exec format error`

martadams89 commented:
> > I've just sync'd my fork to the latest code, updated the workflow to mirror the latest in `main` and everything seemed fine - https://github.com/martadams89/sl-app/actions/runs/12846003655
> > Here's the latest arm64 image - https://registry.hub.docker.com/r/martadams89/sl-app
> 
> @martadams89 Unfortunately this latest build does not seem to work for me. The database migration service fails with errors in the log of:
> 
> ```
> exec /code/.venv/bin/alembic: exec format error
> exec /code/.venv/bin/alembic: exec format error
> exec /code/.venv/bin/alembic: exec format error
> exec /code/.venv/bin/alembic: exec format error
> exec /code/.venv/bin/alembic: exec format error
> ```
> 
> If I disable the migration service, simplelogin-init fails with a single error in the log of:
> 
> `exec /code/.venv/bin/python: exec format error`
> 
> I notice that the file sizes of both the AMD and ARM builds are the same on [your docker hub](https://registry.hub.docker.com/r/martadams89/sl-app/tags) which might mean they have not built correctly?

Looks like something in the move to rye causes the arm64 images to fall over.

I've created a branch/tag of 4.62.0 where arm64 still seems compatible. https://hub.docker.com/repository/docker/martadams89/sl-app/tags/v4-62-0/sha256:74286061a6eaca8163a6c5de2ae49686971ee52f5d05514989bea9d2357f4d99

ctrl-i commented:
> Looks like something in the move to rye causes the arm64 images to fall over.
> 
> I've created a branch/tag of 4.62.0 where arm64 still seems compatible. 

@martadams89 I have just pulled your 4.62.0 branch and can confirm everything works as normal again on that branch.  Hopefully you can resolve the rye dependency.

martadams89 commented:
@ctrl-i I found that amd64 was hardcoded in the Dockerfile and the build workflow was using the rye action with a hardcoded checksum.

I've adjusted the Dockerfile to be image agnostic, and to download the correct version of rye based upon the `TARGETARCH`. The rye checksum is now validated dependant on the architecture of the build server. The build workflow removes the static checksum entry on the rye action, as I don't think this can be called dynamically based upon the architecture of the build server.

my `master` image https://hub.docker.com/repository/docker/martadams89/sl-app/tags has been updated to this version. 

I'd appreciate If @acasajus or another maintainer could review and merge these changes back into the `simple-login:master` branch.

ctrl-i commented:
@martadams89 I tried the new build but I am still getting the 'exec format error' and had to revert again to your 4.62.0 build.

martadams89 commented:
@ctrl-i  I've just double checked here and I'm using my `master` arm64 image - https://hub.docker.com/layers/martadams89/sl-app/master/images/sha256-bc87c0bbd02c5d915b04af1877a695f73bf5a10b499bcc4f4047c6fc7668024a 

you may need to pin to the sha or delete your containers and prune the image.

It looks like the upstream images are smaller, I assume that the switch to rye resulted in smaller image sizes.

ctrl-i commented:
> you may need to pin to the sha or delete your containers and prune the image.

@martadams89 You were right, a quick prune this morning got things moving :)

@acasajus @cquintana92 @nguyenkims Can we get this merged now?  It has been several months now and @martadams89 has shown there is no real reason to not add the arm64 builds back in.

acasajus commented:
Hey, thanks for creating the PR. can you fix the conflict? I'll merge afterwards.

martadams89 commented:
Looks like the build is failing due to the package `memory-profiler` and it's downstream dependancy.

Not sure if this can be updated to a newer version? Then I should be able to create a new build and confirm it works

```
442.1       [stderr]
442.1       warning: build_py: byte-compiling is disabled, skipping.
442.1 
442.1       error: command '/usr/bin/aarch64-linux-gnu-gcc' failed with exit code
442.1       -11
442.1 
442.1       hint: This usually indicates a problem with the package or the build
442.1       environment.
442.1   help: `psutil` (v5.7.2) was included because `simplelogin` (v0.1.0)
442.1         depends on `memory-profiler` (v0.57.0) which depends on `psutil
```

ctrl-i commented:
@martadams89 Great work Martin!  Lets hope this gets merged as promised now.

martadams89 commented:
I'm now happy for this to be merged @acasajus 

I ended up adjusting the Dockerfile to not run uv test as this would intermittently fail when building the image via the docker build action in the workflow.

It would appear to fail without reason, no matter how I segregated the builds it would *occasionally* look to timeout and fail.

If uv checks fail in amd64 then this would fail the whole build anyway and no images should be created, so this should still satisfy the integrity of uv checks.

I've also split the build into a matrix so the amd64 and arm64 builds can run separately, but concurrently. amd64 builds run on amd64 runners, and arm64 builds run on arm64 runners. The built images are then merged in a separate step following the recommended configuration in https://docs.docker.com/build/ci/github-actions/multi-platform/#distribute-build-across-multiple-runners

Finally, I split the release sections out into their own step that depends on a successful build.

Successful builds can be found here https://hub.docker.com/r/martadams89/sl-app/tags

Successful action logs can be found here https://github.com/martadams89/sl-app/actions/runs/12968230283 and https://github.com/martadams89/sl-app/actions/runs/12968163845

martadams89 commented:
I've done multiple tests on my own repo for this and I can't seem to get consistent results. Sometimes the arm builds will seem to fail/timeout when running uv, sometimes they seem to complete - there's no real logic I can follow as to why, amd64 builds complete consistently and quickly - but no matter which way I've tried (matrix, non matrix, split builds, no qemu) I can't seem to get consistent results one way or another. The same workflow will build an arm image (in about 10 minutes, compared to 2 minutes on amd64) and then on a re-run it will seem to fail/timeout. Possibly because I'm running builds back-to-back? I'm not too sure.

It certainly seemed quicker and more consistent on rye.

I think the code in this PR is good and works as expected, I'm just not 100% confident yet that the workflow will run on GitHub Hosted runners 100% of the time..

For what it's worth, the changes I've made is, I've bumped uv to 0.5.24 in both the Dockerfile and the workflow.

In the Dockerfile a hash is now supplied for the amd and arm version of uv, and then based upon the target architecture the relevant binary is downloaded, checksum'd and ran.

On the workflow, I've split the build into a matrix so the amd64 and arm64 builds can run separately, but concurrently.  The built images are then merged in a separate step following the recommended configuration in https://docs.docker.com/build/ci/github-actions/multi-platform/#distribute-build-across-multiple-runners. The repository is now defined as an environment var at the top of the workflow
```
env:
  DOCKERHUB_REPO: simplelogin/app-ci
```

Finally, I split the release sections out into their own step that depends on a successful build.

Successful builds can be found here https://hub.docker.com/r/martadams89/sl-app/tags

Successful action logs can be found here https://github.com/martadams89/sl-app/actions/runs/12969498582/job/36173900431

acasajus commented:
I've tested it and it's quite flaky yet. On the tests I've made it failed 2 out of 3 times. I'm happy to merge if we make it that it can fail. Otherwise I don't want to compromise the amd64 build if the arm one fails. Especially If we have to release quickly. Can we make it so that the amd64 build runs, publishes and so on, and the arm builds in an independent workflow that can fail?

acasajus commented:
I mean, instead of just making it fallible. Right now amd64 waits to on arm to build and it may take a looong time. I would keep the amd64 workflow as it is and have a separate job for arm that pushes the artifacts and images if they manage to build but independently from the amd64 one.

acasajus commented:
Unless I'm mistaken, Docker should be able to have latest pointing to different builds for arm and for amd64. If they have independent jobs and one doesn't update the latest tag, docker will keep the latest for arm and update the amd64. 
Currently the merge step will end up in the same situation just that it will have to wait until arm fails.

martadams89 commented:
> Unless I'm mistaken, Docker should be able to have latest pointing to different builds for arm and for amd64. If they have independent jobs and one doesn't update the latest tag, docker will keep the latest for arm and update the amd64. 
> 

Wouldn't it be whichever image is published most recently would pick up the latest tag? Looking at dockers documentation [here](https://docs.docker.com/build/ci/github-actions/multi-platform/) the examples show the builds in the same workflow. I could be wrong though as I'm not using the "latest" tag.

This [build](https://github.com/martadams89/sl-app/actions/runs/12989893505) produced this [image](https://hub.docker.com/layers/martadams89/sl-app/uv-dev/images/sha256-ae9bff01ba8a759f30bac380ac46b98c5f4c33b968f5e30bd6927a1434ae0ba1)
It was built with the uv-test flag and only produced an amd64 image. Previously there was an amd64 and arm64 image, but after the latest run the old arm64 image got removed.

acasajus commented:
That's weird. Pushing from the cli you can push different platforms with the same tag and it doesn't overlap the previous one. The gh action must be doing some clearing behind the scenes to do that..

martadams89 commented:
@acasajus are you happy to merge in its current state or do you want it refactoring and the workflows splitting out (if this even works correctly)?

acasajus commented:
I need to try out if there's a way to push diferent archs without overwriting the other so the workflows can be independent.

martadams89 commented:
I've updated the Dockerfile to run
`--no-install-package pyre2 --no-install-package pycryptodome`

This was taken from this PR over on linuxserver.io

https://github.com/linuxserver-labs/docker-simplelogin/pull/4

Hopefully this should provide more stable arm64 builds. 

Ideally it would be nice to use the matrix workflow and have one image (simplelogin/app-ci) that is arch agnostic, but having separate images (SimpleLogin/app-ci-arm64) might settle any nerves about needing a quick deployment to your prod environment and having the dependency of an image being built for an architecture you don't internally use..

quietsy commented:
After bashing my head on the arm build inconsistency issue, I finally found the cause: https://github.com/actions/runner-images/issues/11471
You can fix it by changing `runs-on: ubuntu-latest` to `runs-on: ubuntu-22.04` in the workflows.

martadams89 commented:
thanks @quietsy for figuring it out.

@acasajus I've simplifed the workflow so that it does the same steps as present, just with including `linux/arm64`

I tested it [here](https://github.com/martadams89/sl-app/actions/runs/13153211328/job/36704665828) and it built fine.

martadams89 commented:
@acasajus any movement here? I'm now repeatedly successfully building arm64 images

martadams89 commented:
I had to update my GitHub actions file to match the recent changes upstream, along with adjusting the `Dockerfile` (it would't build for some reason).

I took inspiration from @quietsy's `Dockerfile` on the LSIO repo - https://github.com/linuxserver-labs/docker-simplelogin - but it looks like it's building correctly now.

I've merged them changes back into this branch so it should be good to go.

ctrl-i commented:
> I've merged them changes back into this branch so it should be good to go.

I can confirm it is building properly and have installed it successfully.

quietsy commented:
Great job @martadams89 , hope it'll get merged soon.

martadams89 commented:
Any further updates here? @acasajus @cquintana92 @nguyenkims

martadams89 commented:
@luisfalvarez-dev I've tried to get this merged a few times but it seems there's little interest. 

FWIW I build arm64 images on my fork as soon as new updates are released here. I'd love for it to be officially supported, but until then feel free to use my auto-updating fork.
